### ISSUE TITLE
Implement QMK's TAPPING_FORCE_HOLD (and a way to disable it)

### ISSUE THREAD

**Issue by Rougemoot** [david-janssen/kmonad#163]
In kmonad the default behaviour when you tap then hold a `tap-hold` key is the same as when `TAPPING_FORCE_HOLD` is enabled in QMK, which is to activate the `hold` button on the second (held) press.

It would be nice if this was an option one could enable/disable, so that you could have the opposite behaviour instead: on the second (held) press, kmonad would send a press of the `tap` button, allowing it to be auto-repeated.

This is especially important in software like VIM, using home row modifiers. With something like `(defalias jmet (tap-hold-next-release 200 j rmet))` I'm unable to hold `j` to jump down some lines.

**Comment by slotThe** [comment]
This sounds useful; do you know how QMK handles the timeout of this?  I don't assume that you can tap, wait a minute and then still hold down the keys to get a repeat; that sounds like it would result in a lot of unwanted behaviour.

My first instinct is to let the user define a timeout for this; but since this seems to be a solved problem it may be better to just do what QMK already does.

**Comment by Rougemoot** [comment]
QMK just uses the same tapping term as defined for mod-taps. By default in QMK you just define a Tapping Term that'll be used by all mod taps, and a couple other things like tapping force hold, etc.

If you want to, you can define it per key or per function, as you do in kmonad, but yeah, this is where kmonad and QMK differ a bit; in kmonad you always define the tapping term per button, in QMK it's global *unless you want it not to be*

**Comment by slotThe** [comment]
On Fri, Jan 01 2021 04:46, Rougemoot wrote:
> QMK just uses the same tapping term as defined for mod-taps. By
> default in QMK you just define a Tapping Term that'll be used by all
> mod taps, and a couple other things like tapping force hold, etc.
>
> If you want to, you can define it per key or per function, as you do
> in kmonad, but yeah, this is where kmonad and QMK differ a bit; in
> kmonad you always define the tapping term per button, in QMK it's
> global *unless you want it not to be*

Aha I see, so for us it would make sense to let the user insert their
own delay there; thanks!

**Comment by Rougemoot** [comment]
Either insert their own global delay with the option in defcfg or just use the delay provided in the defalias for each button

**Comment by slotThe** [comment]
On Fri, Jan 01 2021 05:30, Rougemoot wrote:
> Either insert their own global delay with the option in defcfg or just
> use the delay provided in the defalias for each button

The former is the only thing we can use, as some buttons (like tap-next)
do not have a delay at all

**Comment by slotThe** [comment]
This functionality is implemented in #168, in case anyone wants to try it

**Comment by Rougemoot** [comment]
I merged this PR onto my local branch of kmonad and compiled it, now I can't use it, whether or not I add the new option in my config. Even with the same exact config as before (which does still work with the old binary), I get this error:

```
kmonad: Parse error at 93:68:
   |
93 |                                 _    brdn brup lp   mctl bldn blup prev pp   next mute vold volu
   |                                                                    ^
unexpected 'p'
expecting ')'
```

which doesn't help much.

You can find my whole config [here](https://github.com/Rougemoot/dotfiles/blob/master/kmonad/applekeeb.kbd).

**Comment by slotThe** [comment]
That's... quite impossible, as I didn't touch that part of the parser at all :)  

This seems to be happening to you because `master` and `develop` have at this point diverged enough to make situations like this possible.  The alias `prev` for `KeyPreviousSong` is present in `master`, but not in `develop` and hence your config doesn't compile.

**Comment by Rougemoot** [comment]
Right so I should be careful to merge your PR onto the master branch then?

**Comment by slotThe** [comment]
> Right so I should be careful to merge your PR onto the master branch then?

That sounds painful to do, but if you're up for it go ahead.  I think the easiest solution would be to make your config compile with `develop` by removing `prev` and whatever else might crop up

**Comment by Rougemoot** [comment]
Ok so I managed to use my config with the develop branch with your PR added, by removing a couple keycodes that weren't doing well with that, as you said.

Now for some reason I can't use my `a` `;` `c` or `m` keys at all, which you can see in my config are `tap_hold_next_release` buttons

**Comment by slotThe** [comment]
On Sun, Jan 03 2021 05:09, Rougemoot wrote:
> Ok so I managed to use my config with the develop branch with your PR
> added, by removing a couple keycodes that weren't doing well with
> that, as you said.
>
> Now for some reason I can't use my `a` `;` `c` or `m` keys at all,
> which you can see in my config are `tap_hold_next_release` buttons

`tap-hold-next-release' buttons work fine for me with this change; does
your layout still match your keyboard after removing those keys?

**Comment by Rougemoot** [comment]
Yeah I just replaced the problematic media keys with `_`s

**Comment by Rougemoot** [comment]
I just realised that those `tap-hold-next-release` keys had a tapping term smaller than the one I used for the `tap-repeat` option. Could that be a clue why I can't use the `tap` portion of those buttons? They go to the hold part right away.

**Comment by slotThe** [comment]
Can you start kmonad with the `-ldebug' option and tell me what you see
when pressing the non-working keys?

**Comment by Rougemoot** [comment]
Here's the debug window showing when I press `p` (which is a default button and works fine) and then `a` (which isn't and doesn't).

**Comment by slotThe** [comment]
On Sun, Jan 03 2021 05:30, Rougemoot wrote:
> I just realised that those `tap-hold-next-release` keys had a tapping
> term smaller than the one I used for the `tap-repeat` option. Could
> that be a clue why I can't use the `tap` portion of those buttons?
> They go to the hold part right away.

No that doesn't matter, the `tap-repeat' part is engaged as soon as we
would normally register a tap, i.e. the timer doesn't start until the
key was released again (at which point the times hook of the button
would be void). 

On Sun, Jan 03 2021 05:36, Rougemoot wrote:
> Here's the debug window showing when I press `p` (which is a default
> button and works fine) and then `a` (which isn't and doesn't).

These look like you are pressing `a', holding it down for 150ms so that
it becomes `sft', and then releasing it again (I will admit that the
output on an apple device appears to be extremely noisy, so I might have
misunderstood some part); can you describe exactly what you are doing,
what you expect to happen and what behaviour you are actually getting?


@MaxGyver83 (sorry for pinging you out of the blue but I couldn't think
of anyone else) can you perhaps try these changes on a non-apple system
and confirm whether or not they work for you?

**Comment by Rougemoot** [comment]
>  (I will admit that the output on an apple device appears to be extremely noisy, so I might have misunderstood some part)

Yeah it's not usually the case, the output is usually much more concise, I think just like you'd expect.

And no, I'm actually tapping `a` just like normal, under 150ms, which is why it's weird that kmonad thinks it should send a `sft` press

**Comment by slotThe** [comment]
On Sun, Jan 03 2021 10:13, Rougemoot wrote:
> And no, I'm actually tapping `a` just like normal, under 150ms, which
> is why it's weird that kmonad thinks it should send a `sft` press

Can you try whether any of this is happening on the "normal" develop
branch, i.e. without the patch, as well?

**Comment by Rougemoot** [comment]
Yep, same exact same thing happens with a fresh clone on the develop branch

**Comment by MaxGyver83** [comment]
> ... @MaxGyver83 (sorry for pinging you out of the blue but I couldn't think of anyone else) can you perhaps try these changes on a non-apple system and confirm whether or not they work for you?

Sure! When I test your branch (before changing anything in my config), at first, everything seems to be quite normal. But at some point, the keyboard becomes unresponsive. The debug output looks like this:

```
--------------------------------------------------------------------------------
Received event: Release <f>
Running hooks
Storing event, current store:
 - Release <f>
 - Release <d>
 - Release <a>
 - Release <s>
 - Press <f>
 - Press <d>
 - Press <s>
 - Press <a>
 - Release <f>
 - Release <d>
 - Release <a>
 - Release <s>
 - Press <f>
 - Press <d>
 - Press <s>
 - Press <a>
 - Release <f>
 - Release <d>
 - Release <a>
 - Release <s>
 - Press <f>
 - Press <d>
 - Press <s>
 - Press <a>
 - Release <f>
 - Release <d>
 - Release <a>
 - Release <s>
 - Press <f>
 - Press <d>
 - Press <a>
 - Press <s>
 - Release <f>
 - Release <d>
 - Release <s>
 - Release <a>
 - Press <f>
 - Press <d>
 - Press <s>
 - Press <a>
 - Release <j>
 - Press <j>
 - Release <j>
 - Release <g>
 - Press <g>
 - Press <j>
 - Release <k>
 - Release <a>
 - Press <a>
 - Press <k>
...
```

I'm not sure when exactly this happens.

I can test the new feature, hoping that my keyboard does not freeze so quickly.

**Comment by MaxGyver83** [comment]
Maybe I've tested it the wrong way!? If I got it right, I have to add this to my config:

```clojure
(defcfg
  input ...
  output ...
  ...
  tap-repeat 5
)
```

Is the number after `tap-repeat` a time in milliseconds? Or do I have to tap 5 times for repetition?

**Comment by slotThe** [comment]
On Sun, Jan 03 2021 10:29, Rougemoot wrote:
> Yep, same exact same thing happens with a fresh clone on the develop
> branch

Alright, that makes more sense.  Since develop only has 16 commits so
far, only one of which is MacOS related, it's probably something that's
been fixed in master.  Because there are some weird `reserved' and
`error' keys in your code, I'm guessing it's due to #55, which was
merged into master in #93.

You can, I guess, try to manually apply that patch and see if it helps.
If it doesn't then only someone with an apple product can really bisect
this.  We definitely don't want to merge master into develop (it *is*
develop after all, things aren't necessarily guaranteed to work ;)

**Comment by MaxGyver83** [comment]
> Is the number after tap-repeat a time in milliseconds? Or do I have to tap 5 times for repetition?

I guess it is a time in milliseconds. I changed it to 1000 and now it works! :+1:

**Comment by MaxGyver83** [comment]
But the problem with the freezing keyboard remains.

**Comment by slotThe** [comment]
On Sun, Jan 03 2021 10:33, Max Schillinger wrote:
> Sure! When I test your branch (before changing anything in my config),
> at first, everything seems to be quite normal. But at some point, the
> keyboard becomes unresponsive. The debug output looks like this:
>
> [...]

Mh interesting; that means we're not releasing a hold when we should.
Probably an easy fix when we know exactly what's causing it.  Do you
perhaps know what tap-* modifier is causing this?  I've tested it
with tap-hold, tap-next, tap-hold-next, and tap-hold-next-release but
maybe I've missed some other modifier.

On Sun, Jan 03 2021 10:39, Max Schillinger wrote:
> Maybe I've tested it the wrong way!? If I got it right, I have to add
> this to my config:
>
> ```clojure
> (defcfg
>   input ...
>   output ...
>   ...
>   tap-repeat 5
> )
> ```
>
> Is the number after `tap-repeat` a time in milliseconds? Or do I have
> to tap 5 times for repetition?

Yes it's a time in milliseconds, i.e. if you tap the key and then,
within the specified time, hold it down again it should press the tap
key.  I've found 100 or 150 to be a decent time for this.

**Comment by MaxGyver83** [comment]
> Mh interesting; that means we're not releasing a hold when we should. Probably an easy fix when we know exactly what's causing it. Do you perhaps know what tap-* modifier is causing this? I've tested it with tap-hold, tap-next, tap-hold-next, and tap-hold-next-release but maybe I've missed some other modifier.

Interestingly, it happens after pressing `u` (the key labeled `r` in QWERTY) which uses no `tap-...` at all. This is my [config](https://gist.github.com/MaxGyver83/84d53866052b13b276b23baaa7efdaf2).

> Yes it's a time in milliseconds, i.e. if you tap the key and then, within the specified time, hold it down again it should press the tap key. I've found 100 or 150 to be a decent time for this.

I noticed a side effect using 1000 ms: Key inputs are being held back by this time. At least for `tap-hold-next-release` keys. Probably no problem for 100 ms.

**Comment by MaxGyver83** [comment]
Some more tests: Now `u` seemed to okay but `o` made the keyboard freeze. `o` is actually `@o` which is an alias: `o (tap-hold 300 o ;)` (And `;` is `ö` because I use `setxkbmap de`.)

**Comment by slotThe** [comment]
> Some more tests: Now `u` seemed to okay but `o` made the keyboard freeze. `o` is actually `@o` which is an alias: `o (tap-hold 300 o ;)` (And `;` is `ö` because I use `setxkbmap de`.)
> 
> EDIT: This time I used `tap-repeat 200`.

Alright, thanks!  I managed to single out the issue and pushed a fix; can try again?

> I noticed a side effect using 1000 ms: Key inputs are being held back by this time. At least for `tap-hold-next-release` keys. Probably no problem for 100 ms.

Incidentally, this should also work "correctly" now (we don't actually need to hold event processing here)

**Comment by MaxGyver83** [comment]
Awesome! Now it works.

But I found a new problem (only in your branch). When I press/hold the `<` key (`102d`) several times, KMonad crashes:

```
Received event: Press <f>
Running hooks
Registering untimed hook: 19
Emitting: Press <i>
Tried cancelling expired hook: 18
Tried cancelling expired hook: 16

--------------------------------------------------------------------------------
Received event: Release <f>
Running hooks
Emitting: Release <i>

--------------------------------------------------------------------------------
Received event: Press <102d>
Running hooks
Block level set to: 1
Registering untimed hook: 21
Registering 500ms hook: 20

--------------------------------------------------------------------------------
Received event: Release <102d>
Running hooks
Unblocking input stream, no stored events
Pushed layer to stack: seldom
 1. seldom
Base-layer: level1

Registering untimed hook: 22
Registering 200ms hook: 23

--------------------------------------------------------------------------------
Received event: Press <102d>
Running hooks
Pushed layer to stack: seldom
 1. seldom
Base-layer: level1

Registering 0ms hook: 24
Registering untimed hook: 25
Registering untimed hook: 26
Cancelling hook: 24
Popped layer from stack: seldom
Base-layer: level1

Tried cancelling expired hook: 23
Tried cancelling expired hook: 20

--------------------------------------------------------------------------------
Received event: Release <102d>
Running hooks
Registering untimed hook: 27

--------------------------------------------------------------------------------
Received event: Press <102d>
Running hooks
Block level set to: 1
Registering 0ms hook: 28
Registering untimed hook: 30
Registering 500ms hook: 29
Cancelling hook: 28
Closing process: emitter_proc
Closing KeySource
Releasing ioctl grab
Closing KeySink
Unregistering Uinput device
Closing Uinput device file
kmonad: LayerNotOnStack "seldom"
```

**Comment by MaxGyver83** [comment]
Another issue only with your branch:

When I type quickly `f`, `j`, `f` (mapped to `i`, `t`, `i` in my layout), I get `i`, `i`, `t`.

**Comment by slotThe** [comment]
Thanks; the crash is fixed, the other issue needs some thinking...  It's a very weird interaction between two `tap-hold-next-release` style buttons

**Comment by MaxGyver83** [comment]
I tested again and it crashed again :cry: :

```
Opening KeySink
Registering Uinput device
Running UinputSink command: "/bin/sleep 1 && setxkbmap de -option compose:ralt && xset r rate 300 50"
Opening KeySource
Initiating ioctl grab
Launching process: emitter_proc

--------------------------------------------------------------------------------
Received event: Press <102d>
Running hooks
Block level set to: 1
Registering untimed hook: 2
Registering 500ms hook: 1

--------------------------------------------------------------------------------
Received event: Release <102d>
Running hooks
Pushed layer to stack: seldom
 1. seldom
Base-layer: level1

Registering untimed hook: 3
Unblocking input stream, no stored events
Registering 200ms hook: 4

--------------------------------------------------------------------------------
Received event: Press <102d>
Running hooks
Registering 0ms hook: 6
Registering 0ms hook: 5
Cancelling hook: 6
Pushed layer to stack: seldom
 1. seldom
Base-layer: level1

Registering untimed hook: 7
Registering untimed hook: 8
Cancelling hook: 5
Popped layer from stack: seldom
Base-layer: level1


--------------------------------------------------------------------------------
Received event: Release <102d>
Running hooks
Registering untimed hook: 9
Tried cancelling expired hook: 4

--------------------------------------------------------------------------------
Received event: Press <102d>
Running hooks
Block level set to: 1
Registering 0ms hook: 10
Registering untimed hook: 12
Registering 500ms hook: 11
Cancelling hook: 10
Closing process: emitter_proc
Closing KeySource
Releasing ioctl grab
Closing KeySink
Unregistering Uinput device
Closing Uinput device file
kmonad: LayerNotOnStack "seldom"
```

**Comment by slotThe** [comment]
On Wed, Jan 06 2021 01:06, Max Schillinger wrote:
> I tested again and it crashed again :cry: :

Keyboards are weird.  It seems that one condition is outracing the
other, causing a layer to be pushed first and then immediately popped,
instead of the other way around.  Apparently, this just didn't happen on
my laptop when testing :o

I pushed a hack that fixes (fingers crossed) this; instead of hoping
that the events happen in the right order we now simply wait a little
(5ms, should be imperceptible) before pressing the button again.  I have
mashed the key really hard and couldn't reproduce this crash, so
hopefully this actually works now

**Comment by MaxGyver83** [comment]
Cool! I've tested it again and had no crashes.

**Comment by MaxGyver83** [comment]
@slotThe : Are you still working on this? Is it difficult to get the keys emitted in the right order?

**Comment by slotThe** [comment]
On Thu, Jan 28 2021 11:58, Max Schillinger wrote:
> @slotThe : Are you still working on this? Is it difficult to get the
> keys emitted in the right order?

I am still working on this, but I've sadly been swamped with other
duties lately.  It is actually quite tricky to get this to work, yes :(

I don't remember the exact details (but I have written them down
somewhere!), but the gist of it is that, since we're creating another
hook that waits for a keypress once any tap-style button is pressed,
this can't properly be "held" in the same way a "normal" button press
can when we hold event-processing when pressing a `tap-hold' style
button afterwards.

From what I gathered the last time I looked into it, this probably means
that I need a completely new design for how to implement this feature—I
definitely want this behaviour to be part of kmonad, it's quite
convenient, I just can't promise _when_ it'll hit.

**Comment by david-janssen** [comment]
Hey everyone,

What great efforts :-)! I really like idea. I am putting this in 'middle-of-pile'. Looking at some of the other bugs and feature requests makes me think that some redesign on the core is required to more easily support certain features that never occured to me when I first wrote KMonad. I'm going to have a go at that redesign first, and then implement this feature using the changes to the core. Seeing as my time is very limited this isn't going to be any time soon, but I completely intend to be the tortoise in this race.

**Comment by abdalrahman-ali** [comment]
Hi @slotThe, I tried your fix but didn't work for me. Using windows, not sure if this matters...
I'm using home row mods like:
(defalias jmet (tap-hold-next-release 200 j rmet))
Any Ideas why it wouldn't work?

**Comment by jokesper** [comment]
I think this may be possible when using #882 in combination with `tap-hold` style buttons:
```clojure
(tap-hold 200 (retap 200 a a) b)
```
